### ISSUE TITLE
Operator spec of Split operator's attribute num_outputs is wrong

### ISSUE THREAD

**Issue by zhenhuaw-me** [onnx/onnx#5766]
## The Problem

[In op version 18](https://onnx.ai/onnx/operators/text_diff_Split_13_18.html), we added the spec of Split attribute `num_outputs` with the statement

> **num_outputs - INT** :   Number of outputs to split parts of the tensor into. If the tensor is not evenly splittable the last chunk will be smaller.

Consider the case `Split(input, 4)` where `input` has dim `[5]`, what the shape of the outputs will be?

According to the [shape inference code](https://github.com/onnx/onnx/blob/b60f69412abb5393ab819b936b473f83867f6c87/onnx/defs/tensor/defs.cc#L780C1-L783C50), the shape of the output tensors should be `[2], [2], [2], [-1]` which is illegal.

```c++
                int chunk_size = (split_dim_value / num_outputs) + 1;
                int last_chunk_size = split_dim_value - (chunk_size * (num_outputs - 1));
                split.resize(num_outputs - 1, chunk_size);
                split.push_back(last_chunk_size);
```

## Reference

It seems to me that we were trying to combine the definition of numpy/TF and Torch.

For [**Numpy**](https://numpy.org/doc/stable/reference/generated/numpy.split.html)/[**TF**](https://www.tensorflow.org/api_docs/python/tf/split), they use the similar semantic, but REJECT if not divisible:

> If num_or_size_splits is an int, then it splits value along the dimension axis into num_or_size_splits smaller tensors. This requires that value.shape[axis] is divisible by num_or_size_splits.

[**Torch**](https://pytorch.org/docs/stable/generated/torch.split.html) instead declares it as a _split size_, in which way indivisible is acceptable:

> If split_size_or_sections is an integer type, then [tensor](https://pytorch.org/docs/stable/generated/torch.tensor.html#torch.tensor) will be split into equally sized chunks (if possible). Last chunk will be smaller if the tensor size along the given dimension dim is not divisible by split_size.

## Possible Solutions

1. Keep the `num_outputs` and reject if indivisible.
2. Change to Torch's definition, thus we need to change the attribute name.

**Comment by zhenhuaw-me** [comment]
The Split-18 PR: https://github.com/onnx/onnx/pull/4481 @p-wysocki 

What's the intention of `num_outputs`, as it is different from both Torch and TensorFlow.

cc @onnx/sig-operators-approvers

**Comment by liqunfu** [comment]
Good catch! The onnx spec need to be updated to require the input being evenly splittable by num_outputs, if num_outputs is specified. This is the same as tf: "that value.shape[axis] is divisible by num_or_size_splits". In the case where num_or_size_splits in TF and split_size_or_sections in pytorch NOT being a single integer, it is covered by "split" input in ONNX. To support split_size_or_sections being a single integer in Pytorch, we need add an attribute "split_size" where the last output may be smaller than split_size. 

Here is how cases in TF and Pytroch are covered in ONNX:
Pytorch:
input = [1, 2, 3, 4, 5], split_size_or_sections = 2 => output = ([1, 2], [3, 4], [5]); // ONNX split_size = 2
input = [1, 2, 3, 4, 5], split_size_or_sections = [2, 3] => output = ([1, 2], [3, 4, 5]) // ONNX split = [2, 3]
TF:
input = [1, 2, 3, 4, 5], num_or_size_splits = 2 => no dividable error; // ONNX no dividable error
input = [1, 2, 3, 4, 5, 6], num_or_size_splits = 2 => output = ([0, 1, 2], [3, 4, 5]; // ONNX num_outputs = 2
input = [1, 2, 3, 4, 5], split_size_or_sections = [2, 3] => output = ([1, 2], [3, 4, 5]) // ONNX split = [2, 3]

**Comment by zhenhuaw-me** [comment]
@liqunfu Thank you for the update! That sounds good. Do we plan to include this in next release?

**Comment by zhenhuaw-me** [comment]
Copy https://github.com/onnx/onnx/issues/4742#issuecomment-1839934443

One step further, we can fix the current attribute `num_outputs` by aligning it with [`numpy.array_split`](https://numpy.org/doc/stable/reference/generated/numpy.array_split.html) and [`torch.tensor_split`](https://pytorch.org/docs/stable/generated/torch.tensor_split.html)

> If indices_or_sections is an integer n or a zero dimensional long tensor with value n, input is split into n sections along dimension dim. If input is divisible by n along dimension dim, each section will be of equal size, input.size(dim) / n. If input is not divisible by n, the sizes of the first int(input.size(dim) % n) sections will have size int(input.size(dim) / n) + 1, and the rest will have size int(input.size(dim) / n).
